src/ReplaceReferencesCapableTrait.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+namespace Dhii\Config;
+
+use InvalidArgumentException;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\ContainerInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Psr\Container\NotFoundExceptionInterface;
+
+/**
+ * Trait ReplaceReferencesCapableTrait.
+ *
+ * Adds ability to replace all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+ *
+ * @since [*next-version*]
+ */
+trait ReplaceReferencesCapableTrait
+{
+    /**
+     * Replaces all tokens in the form `${key}` in a string with corresponding values retrieved from a container.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|Stringable  $input          Input string to find and replace references
+     * @param ContainerInterface $container
+     * @param string|Stringable  $startDelimiter
+     * @param mixed              $default        String to replace reference if key not found in container.
+     * @param string|Stringable  $endDelimiter
+     *
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     *
+     * @return string The resulting string.
+     */
+    protected function _replaceReferences($input, ContainerInterface $container, $default = null, $startDelimiter = '${', $endDelimiter = '}')
+    {
+        $input = $this->_normalizeString($input);
+
+        $startDelimiter = $this->_normalizeRegexpDelimiter($startDelimiter);
```

Comment:
There is no need to escape every character. You know the RegEx delimiter: judging by the below code, it's `/`. Therefore, you can use [`preg_quote()`][] to do what you are doing here. Just normalize to string before ;)

[`preg_quote()`]: http://php.net/manual/en/function.preg-quote.php

Reply:
Cool. I didn't know about this function 🤦‍♂️